src/lib/support/SplitLambda.h

```diff
@@ -0,0 +1,42 @@
+#pragma once
+#include <tuple>
+#include <utility>
+
+namespace chip {
+
+namespace detail {
+template <typename T>
+struct SplittedLambdaCallerImpl;
+
+template <class TReturn, class TLambda, class... TArgs>
+struct SplittedLambdaCallerImpl<TReturn (TLambda::*)(TArgs...) const>
+{
+    using CallSignature = TReturn(TArgs... args, void * context);
+    template <std::size_t I>
+    using TupleElement = std::tuple_element<I, std::tuple<CallSignature *, void *>>;
+
+    static TReturn Call(TArgs... args, void * context) { return (*static_cast<TLambda *>(context))(std::forward<TArgs>(args)...); }
+};
+
+} // namespace detail
+
+template <class TLambda>
+struct SplittedLambda : detail::SplittedLambdaCallerImpl<decltype(&TLambda::operator())>
```

Comment:
all these classes need documentation (including the Shim one): what is their intent, what do they do, what is their usage.